## 📋 Quality Gate Results

| Check | Status |
|-------|--------|
| Knip | ✅ success |
| Lint | ❌ failure |
| Build | ✅ success |
| Infra Tests | ❌ skipped |
| Unit Tests | ❌ skipped |
| Component Tests | ❌ skipped |
| Perf Tests | ❌ skipped |
| Visual Tests | ❌ skipped |


### ❌ Lint Failure Details
```text

> hrm@0.30.0 lint /home/runner/work/hrm/hrm
> eslint app/ components/ constants/ context/ hooks/ lib/ services/ tests/ types/ utils/ server.ts proxy.ts --cache


/home/runner/work/hrm/hrm/tests/unit/services/spotifyPolling.test.ts
  18:37  error  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  30:26  error  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 2 problems (2 errors, 0 warnings)

 ELIFECYCLE  Command failed with exit code 1.
```
### ❌ Infrastructure Test Failure Details
```text
Log file not found.
```
### ❌ Unit Test Failure Details
```text
Log file not found.
```
### ❌ Component Test Failure Details
```text
Log file not found.
```
### ❌ Visual Test Failure Details
```text
Log file not found.
```
### ❌ Performance Test Failure Details
```text
Log file not found.
```
⚠️ **Some checks failed.** Full logs available in [workflow artifacts](https://github.com/arii/hrm/actions/runs/21573804115).

---
> Report generated for commit: `b88dcf02afcb9d14bc5b0075353ab8f0e754fb65`

--- a/services/spotifyPolling.ts
+++ b/services/spotifyPolling.ts
@@ -42,21 +42,6 @@ export class SpotifyPolling implements SpotifyService {
    * Test-only properties for inspecting internal state, defined only in 'test' env.
    * See docs/TYPESCRIPT_PATTERNS.md for more info on this pattern.
    */
-  public _test_ =
-    process.env.NODE_ENV === 'test'
-      ? {
-          getPollInterval: () => this.pollInterval,
-          getTokenRefreshInterval: () => this.tokenRefreshInterval,
-          setPollInterval: (interval: NodeJS.Timeout | null) => {
-            this.pollInterval = interval
-          },
-          setTokenRefreshInterval: (interval: NodeJS.Timeout | null) => {
-            this.tokenRefreshInterval = interval
-          },
-          isEmptyResponseError: this.isEmptyResponseError.bind(this),
-        }
-      : undefined
-
   private readonly broadcastUpdate: (message: ServerMessage) => void
 
   private lastTrackId: string | null = null
@@ -89,6 +74,106 @@ export class SpotifyPolling implements SpotifyService {
       env.SPOTIFY_CLIENT_SECRET
     )
   }
+
+  private getCurrentlyPlaying = async () => {
+    try {
+      if (!this.sdk) {
+        logger.debug('Spotify SDK not initialized, skipping poll')
+        return
+      }
+      const playbackState = await this.sdk.player.getCurrentlyPlayingTrack()
+
+      if (!playbackState || !playbackState.item) {
+        // Nothing playing, 204, or private session
+        if (this.lastPlaybackState !== false) {
+          this.lastPlaybackState = false
+          this.state = {
+            ...this.state,
+            trackId: null,
+            trackName: 'Nothing is currently playing.',
+            artist: '',
+            albumName: '',
+            albumArtUrl: '',
+            isPlaying: false,
+          }
+          this.broadcastUpdate({
+            type: 'SPOTIFY_UPDATE',
+            payload: this.getState(),
+          })
+        }
+        return
+      }
+
+      const item = playbackState.item
+      const isPlaying = playbackState.is_playing
+
+      // Only broadcast if track ID or playback state has changed
+      if (
+        item.id !== this.lastTrackId ||
+        isPlaying !== this.lastPlaybackState
+      ) {
+        this.lastTrackId = item.id
+        this.lastPlaybackState = isPlaying
+
+        const trackName = item.name
+        const trackId = item.id
+        let artistName = ''
+        let albumName = ''
+        let albumArtUrl = ''
+
+        if (item.type === 'track') {
+          const track = item as Track
+          artistName = track.artists.map((a) => a.name).join(', ')
+          albumName = track.album.name
+          albumArtUrl = track.album.images?.[0]?.url ?? ''
+        } else if (item.type === 'episode') {
+          const episode = item as Episode
+          artistName = episode.show.publisher
+          albumName = episode.show.name
+          albumArtUrl = episode.show.images?.[0]?.url ?? ''
+        }
+
+        this.state = {
+          ...this.state,
+          trackId,
+          trackName,
+          artist: artistName,
+          albumName,
+          albumArtUrl,
+          isPlaying,
+        }
+
+        this.broadcastUpdate({
+          type: 'SPOTIFY_UPDATE',
+          payload: this.getState(),
+        })
+      }
+    } catch (error) {
+      await handleSpotifyApiError(error, () => this.checkAndRefreshSdkToken())
+    }
+  }
+
+  /**
+   * @internal
+   * This is a public property for testing purposes.
+   * It is only defined in a test environment.
+   */
+  public _test_ =
+    process.env.NODE_ENV === 'test'
+      ? {
+          getPollInterval: () => this.pollInterval,
+          getTokenRefreshInterval: () => this.tokenRefreshInterval,
+          setPollInterval: (interval: NodeJS.Timeout | null) => {
+            this.pollInterval = interval
+          },
+          setTokenRefreshInterval: (interval: NodeJS.Timeout | null) => {
+            this.tokenRefreshInterval = interval
+          },
+          isEmptyResponseError: this.isEmptyResponseError.bind(this),
+          getSdk: this.getSdk.bind(this),
+          getCurrentlyPlaying: this.getCurrentlyPlaying.bind(this),
+        }
+      : undefined
 
   public static async create(
     broadcastUpdate: (message: ServerMessage) => void
@@ -229,84 +314,6 @@ export class SpotifyPolling implements SpotifyService {
       clearInterval(this.tokenRefreshInterval)
       this.tokenRefreshInterval = null
       logger.debug('Token refresh interval cleared.')
-    }
-  }
-
-  private getCurrentlyPlaying = async () => {
-    try {
-      if (!this.sdk) {
-        logger.debug('Spotify SDK not initialized, skipping poll')
-        return
-      }
-      const playbackState = await this.sdk.player.getCurrentlyPlayingTrack()
-
-      if (!playbackState || !playbackState.item) {
-        // Nothing playing, 204, or private session
-        if (this.lastPlaybackState !== false) {
-          this.lastPlaybackState = false
-          this.state = {
-            ...this.state,
-            trackId: null,
-            trackName: 'Nothing is currently playing.',
-            artist: '',
-            albumName: '',
-            albumArtUrl: '',
-            isPlaying: false,
-          }
-          this.broadcastUpdate({
-            type: 'SPOTIFY_UPDATE',
-            payload: this.getState(),
-          })
-        }
-        return
-      }
-
-      const item = playbackState.item
-      const isPlaying = playbackState.is_playing
-
-      // Only broadcast if track ID or playback state has changed
-      if (
-        item.id !== this.lastTrackId ||
-        isPlaying !== this.lastPlaybackState
-      ) {
-        this.lastTrackId = item.id
-        this.lastPlaybackState = isPlaying
-
-        const trackName = item.name
-        const trackId = item.id
-        let artistName = ''
-        let albumName = ''
-        let albumArtUrl = ''
-
-        if (item.type === 'track') {
-          const track = item as Track
-          artistName = track.artists.map((a) => a.name).join(', ')
-          albumName = track.album.name
-          albumArtUrl = track.album.images?.[0]?.url ?? ''
-        } else if (item.type === 'episode') {
-          const episode = item as Episode
-          artistName = episode.show.publisher
-          albumName = episode.show.name
-          albumArtUrl = episode.show.images?.[0]?.url ?? ''
-        }
-
-        this.state = {
-          ...this.state,
-          trackId,
-          trackName,
-          artist: artistName,
-          albumName,
-          albumArtUrl,
-          isPlaying,
-        }
-
-        this.broadcastUpdate({
-          type: 'SPOTIFY_UPDATE',
-          payload: this.getState(),
-        })
-      }
-    } catch (error) {
-      await handleSpotifyApiError(error, () => this.checkAndRefreshSdkToken())
     }
   }
 

--- a/tests/unit/services/spotifyPolling.test.ts
+++ b/tests/unit/services/spotifyPolling.test.ts
@@ -14,9 +14,7 @@ describe('SpotifyPolling', () => {
   })
 
   it('should throw an error when getSdk is called before initialization', () => {
-    // Directly accessing a private method for testing purposes.
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    expect(() => (spotifyPolling as any).getSdk()).toThrow(
+    expect(() => spotifyPolling._test_?.getSdk()).toThrow(
       'Spotify SDK has not been initialized.'
     )
   })
@@ -28,8 +26,7 @@ describe('SpotifyPolling', () => {
     // Act & Assert
     // The method should complete without throwing an error because of the guard clause.
     await expect(
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      (spotifyPolling as any).getCurrentlyPlaying()
+      spotifyPolling._test_?.getCurrentlyPlaying()
     ).resolves.not.toThrow()
 
     // It should have returned early, so no broadcast should have been sent.

--- a/tests/unit/services/spotifyPolling.test.ts
+++ b/tests/unit/services/spotifyPolling.test.ts
@@ -15,6 +15,7 @@ describe('SpotifyPolling', () => {
 
   it('should throw an error when getSdk is called before initialization', () => {
     // Directly accessing a private method for testing purposes.
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
     expect(() => (spotifyPolling as any).getSdk()).toThrow(
       'Spotify SDK has not been initialized.'
     )
@@ -27,6 +28,7 @@ describe('SpotifyPolling', () => {
     // Act & Assert
     // The method should complete without throwing an error because of the guard clause.
     await expect(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
       (spotifyPolling as any).getCurrentlyPlaying()
     ).resolves.not.toThrow()
 

